design/service-binding.md

```diff
@@ -0,0 +1,356 @@
+# Service Binding Proposal for Strimzi
+
+This document is intended to progress discussion about how to enable Strimzi to work well with the Service Binding Operator. It includes some suggested enhancements to both Strimzi and the Service Binding Operator. It would be good to settle on an agreement about how these two technologies can best work together so we can begin the technical work to deliver.
+
+The Service Binding Operator is responsible for binding services such as databases and message brokers to runtime applications in Kubernetes. It's still under development, but the intention is that it becomes part of Operator Lifecycle Management.
+
+## Service Binding Operator today
+
+Today, Strimzi does not fit very nicely with service binding. With the current Service Binding Opeerator implementation, you need to include a template in the `ServiceBindingRequest` to extract the address information from the listener status.
+
+``` yaml
+apiVersion: apps.openshift.io/v1alpha1
+kind: ServiceBindingRequest
+metadata:
+  name: barista-kafka
+spec:
+  applicationSelector:
+    group: apps
+    resource: deployments
+    resourceRef: barista-kafka
+    version: v1
+  backingServiceSelector:
+    group: kafka.strimzi.io
+    kind: Kafka
+    resourceRef: my-cluster
+    version: v1beta1
+  customEnvVar:
+    - name: KAFKA_BOOTSTRAP_SERVERS
+      value: |-
+        {{- range .status.listeners -}}
+          {{- if and (eq .type "plain") (gt (len .addresses) 0) -}}
+            {{- with (index .addresses 0) -}}
+              {{ .host }}:{{ .port }}
+            {{- end -}}
+          {{- end -}}
+        {{- end -}}
+```
+
+While this is quite a clever use of a Go template, it has several problems.
+
+1) The code is included in an annotation for the user's custom resource. This is fragile and ugly.
+1) The code includes the name of a listener. If a different listener name is being used, the code is incorrect.
+1) The code picks just the first of the array of addresses. It's common Kafka practice to have a list of bootstrap servers.
```

Comment:
In general this is true. But in our case, Strimzi always provides a service which loadbalances between all available brokers. So I do not think this is true with Strimzi.

Reply:
If the intention is for users to always call a single service that load balances across several brokers will there ever be a scenario where a single listener has multiple addresses listed? And if not what was the motivation for having "addresses" as an array in the status object?

Reply:
I actually described it in one of the followup comments:

> The only situation when we actually have multiple addresses for single listener is with the external listener of node port type where the whole situation is a bit more complicated. But I guess the service binding operator works only within the same cluster so this should not be relevant, right?

Reply:
Provided we pick a syntax that is capable of supporting one or more addresses, we can accommodate the usual situation with a single address as well as the unusual situation with more than one. The latest version of the document does that.